gtk/src/subclass/container.rs

```diff
@@ -4,122 +4,136 @@ use glib::translate::*;
 
 use glib::subclass::prelude::*;
 
+use glib::Cast;
+
 use super::widget::WidgetImpl;
 use Container;
 use Widget;
 use WidgetPath;
 
 pub trait ContainerImpl: ContainerImplExt + WidgetImpl {
-    fn add(&self, container: &Container, widget: &Widget) {
+    fn add(&self, container: &Self::Type, widget: &Widget) {
         self.parent_add(container, widget)
     }
 
-    fn remove(&self, container: &Container, widget: &Widget) {
+    fn remove(&self, container: &Self::Type, widget: &Widget) {
         self.parent_remove(container, widget)
     }
 
-    fn check_resize(&self, container: &Container) {
+    fn check_resize(&self, container: &Self::Type) {
         self.parent_check_resize(container)
     }
 
-    fn set_focus_child(&self, container: &Container, widget: Option<&Widget>) {
+    fn set_focus_child(&self, container: &Self::Type, widget: Option<&Widget>) {
         self.parent_set_focus_child(container, widget)
     }
 
-    fn child_type(&self, container: &Container) -> glib::Type {
+    fn child_type(&self, container: &Self::Type) -> glib::Type {
         self.parent_child_type(container)
     }
 
-    fn get_path_for_child(&self, container: &Container, widget: &Widget) -> WidgetPath {
+    fn get_path_for_child(&self, container: &Self::Type, widget: &Widget) -> WidgetPath {
         self.parent_get_path_for_child(container, widget)
     }
 
-    fn forall(&self, container: &Container, include_internals: bool, callback: &Callback) {
+    fn forall(&self, container: &Self::Type, include_internals: bool, callback: &Callback) {
         self.parent_forall(container, include_internals, callback);
     }
 }
 
-pub trait ContainerImplExt {
-    fn parent_add(&self, container: &Container, widget: &Widget);
-    fn parent_remove(&self, container: &Container, widget: &Widget);
-    fn parent_check_resize(&self, container: &Container);
-    fn parent_set_focus_child(&self, container: &Container, widget: Option<&Widget>);
-    fn parent_child_type(&self, container: &Container) -> glib::Type;
-    fn parent_get_path_for_child(&self, container: &Container, widget: &Widget) -> WidgetPath;
-    fn parent_forall(&self, container: &Container, include_internals: bool, callback: &Callback);
+pub trait ContainerImplExt: ObjectSubclass {
+    fn parent_add(&self, container: &Self::Type, widget: &Widget);
```

Comment:
I was wondering if it could make sense to not pass `&Self::Type` at all here as it's kind of redundant but I think it still makes sense:
  * Usually you need it, so that would otherwise be an additional call to `self.get_instance()`
  * Getting it from `&self` involves reference counting: not a usability problem as that's handled automatically in Rust, but it means additional atomic integer operations that could be avoided otherwise

As such, let's keep it like this for now :)

Reply:
Also for people using an IDE or LSP plugin for their editor this comes for free. The signature of the trait functions would be autocompleted so typing the `&Self::Type` is not much of an effort.

Reply:
> As such, let's keep it like this for now :)

Also the soundness issues with `get_instance()` (https://github.com/gtk-rs/gtk-rs/issues/46), which will need some sort of breaking API change to address. I'm personally happy to not be adding a bunch of calls to the function when it can easily be used to produce unsound behavior. And the way it works may have to change to fix the soundness issue anyway.

Reply:
Well, it's difficult to accidentally run into that soundness issue so I wouldn't worry too much about it.

We can also make it easily cause an assertion (or direct segfault because dereferencing some random pointer) if that happens (check that the GType inside the instance is the one we would expect), which should make it even less likely to happen in practice. Want to give that a try?